File: data-structures/ds-stk/cpp/stack.cpp
//About stack
//A stack is a linear data structure that serves as a collection of elements, with following main operations:
//Push operation, which adds an elements to the stack.
//Pop operation, which removes the most recently added element that was not yet removed.

//The push and pop operations occur only at one end of the structure, referred to as the top of the stack.
//The order in which elements come off a stack gives rise to its alternative name, LIFO (for last in, first out).


#include<iostream>
using namespace std;
int A[100],choose,m,n,top,i;
void push();
void pop();
void display();

int main()
{
    top=-1;
    cout<<"Enter the size of stack:: ";
    cin>>m;

    cout<<"\n1.PUSH \n2.POP \n3.DISPLAY \n4.EXIT\n ";

    do                                                 //do-while loop is used to ask the user for their choice
    {
        cout<<"\n\nChoose the number :: ";
        cin>>choose;

        switch(choose)
        {
        case 1:
        {
            push();
            break;
        }

        case 2:
        {
            pop();
            break;
        }

        case 3:
        {
            display();
            break;
        }

        case 4:
        {
            cout<<"EXIT";
            break;
        }

        default:
        {
            cout<<"INVALID INPUT";
        }
        }
    }
    while(choose!=4);
    return 0;
}


void push()                                           //subprogram for operating the push operation in stack
{
    if(top>=m-1)                                        //if top is greater than and equals to size of array-1
    {
        cout<<"Stack overflow";
    }
    else                                                //else push the element
    {
        cout<<"Enter a value to be pushed:: ";
        cin>>n;
        top++;
        A[top]=n;
    }
}

void pop()                           //subprogram for operating the pop operation in stack
{
    if(top<=-1)                          //if top is less than and equals to -1
    {
        cout<<"Stack under flow";
    }
    else                                       //else pop the element
    {
        cout<<"Poped element is "<<A[top];
        top--;
    }
}

void display()                              //subprogram to display the elements of stack
{
    if(top>=0)                                 //if top is greater than and equals to 0
    {
        cout<<"Stack elements are:: ";
        for(i=top; i>=0; i--)
            cout<<"\t"<< A[i];
    }
    else                                              //else print stack is empty
    {
        cout<<"stack is empty.";
    }
}

File: data-structures/ds-stk/python2/stack.py
# Implementing the Stack class


class Stack:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return self.items == []

    def push(self, data):
        self.items.append(data)

    def pop(self):
        return self.items.pop()

    def peek(self):
        return self.items[len(self.items) - 1]

    def size(self):
        return len(self.items)

    def printStack(self):
        print(self.items)

# End of Stack class


def Main():
    # create new Stack
    st = Stack()
    print("#### Welcome ####")
    op = 1

    while op != 7:
        print("1.isEmpty\n2.Push\n3.POP\n4.Peek\n5.size\n6.Print\n7.Exit")
        op = int(input("Select your option : "))

        print("-------------------------------------")
        if(op == 1):
            if(st.isEmpty()):
                print("Stack is Empty")
            else:
                print("Stack is not empty")
        elif(op == 2):
            data = input("Enter data : ")
            st.push(data)
        elif(op == 3):
            if(st.isEmpty()):
                print("Can't pop because stack is empty")
            else:
                print(st.pop(), " is popped out from stack.")
        elif(op == 4):
            if(st.isEmpty()):
                print("No element in stack.")
            else:
                print(st.pop(), " is the top most element.")
        elif(op == 5):
            if(st.size() == 0):
                print("Stack is empty")
            else:
                print("stack have ", st.size(), " element(s).")
        elif(op == 6):
            if(st.isEmpty()):
                print("Stack is empty")
            else:
                print("Stack : ")
                st.printStack()
        elif(op != 7):
            print("Invalid Option.\n")

        print("-------------------------------------")
    print("GoodBye")


if __name__ == "__main__":
    Main()
